Split metrics on last client write

matte21 opened the pull request:
This PR is the first step towards the changes discussed in #111 .
It splits metrics on the latencies of the lifecycles of NetAtts and subnets (NAs) on the basis of the last client precursor.

This PR self-contained and mergeable, but it's partial. It does not take into account in any way the fact that client precursors could be deletions of API objects as well (as pointed out [here](https://github.com/MikeSpreitzer/kube-examples/pull/111#issuecomment-625897528)).  It does not take into account the fact that implementation of API objects could be delayed because controllers could fail. Let's merge this and do remaining things in follow-up PRs.

matte21 committed:
Diversify IPAM controller metrics

matte21 committed:
Add creation timestamp of subnet to NetworkAttachment's status

matte21 committed:
Diversify Connection Agent metrics

matte21 committed:
Better recording of virtual network relevance time

Record virtual network relevance time in a more precise way:

1. When doing a sync with pre-existing network interfaces at start
up, correctly pick the earliest relevance time rather than the one
associated to the first local attachment to be processed.

2. Add a prometheus gauge to track the delay of the recorded
relevance time wrt the real relevance time.

3. Log a warning when a relevance delay is detected.

matte21 commented on staging/kos/pkg/apis/network/types.go:
```diff
@@ -221,6 +221,11 @@ type NetworkAttachmentStatus struct {
+	// SubnetCreationTime is the API server write time of the SubnetSectionSpec
+	// of the subnet identified by NetworkAttachmentSpec.Subnet.
+	// +optional
+	SubnetCreationTime metav1.MicroTime
```
When we add metrics to track delays caused by controllers' downtime we'll want to add a `lastImplementationPrecursor` with the name of the last controller to start and the time of that start. For consistency (and better design) we'll want to get rid of this and have a more generic field like `lastClientWrite` that carries the type of the last client write (e.g. subnet creation, natt creation) and the time of that write. IMO those two fields should go in `ExtendedObjectMeta` rather than `Status`(although they are written by controllers rather than API servers). But for now this serves the intended purpose.

MikeSpreitzer replied:
Yes.

MikeSpreitzer commented on staging/kos/pkg/controllers/connectionagent/connection_agent.go:
```diff
@@ -251,20 +277,27 @@ type ConnectionAgent struct {
 
-	// NASectionImpl to remote network interface creation latency
-	localImplToRemoteIfcHistogram prometheus.Histogram
+	// Seconds from attachment NASectionImpl to creation of remote network
```
This is interesting in only two of the four cases.  I would not even produce observations for the two uninteresting cases.

matte21 replied:
Done in 52413b4 .

MikeSpreitzer commented on staging/kos/pkg/controllers/connectionagent/connection_agent.go:
```diff
@@ -817,13 +892,22 @@ func (ca *ConnectionAgent) syncS2VNState(attNSN k8stypes.NamespacedName, att *ne
+	if att.Status.SubnetCreationTime.After(vnRelevanceTime) {
+		vnRelevanceTrigger = firstLocalNASubnetLastClientWrite
+		vnRelevanceTime = att.Status.SubnetCreationTime.Time
+	}
```
Common code here with `getLastClientWriteInfo`, maybe create a method for it?

matte21 replied:
Done in 52413b4 .

MikeSpreitzer commented on staging/kos/pkg/controllers/connectionagent/connection_agent.go:
```diff
@@ -911,6 +1000,41 @@ func (ca *ConnectionAgent) clearStage1VNState(vni uint32, namespace string) {
+	}
+
+	dt := s1VNS.relevanceTime.Sub(newRelevanceTime).Seconds()
+	if dt > vnRelevanceDelayGraceSecs && dt > s1VNS.relevanceDelaySecs {
```
I wonder if we should threshhold every update, not just the first.  That is, make this test
```
if dt - s1VNS.relevanceDelaySecs > vnRelevanceDelayGraceSecs {
```

matte21 replied:
Done in 52413b4 .

MikeSpreitzer commented on staging/kos/pkg/controllers/connectionagent/connection_agent.go:
```diff
@@ -911,6 +1000,41 @@ func (ca *ConnectionAgent) clearStage1VNState(vni uint32, namespace string) {
+	dt := s1VNS.relevanceTime.Sub(newRelevanceTime).Seconds()
+	if dt > vnRelevanceDelayGraceSecs && dt > s1VNS.relevanceDelaySecs {
+		// Make some noise.
+		klog.Warningf("vni %d: recorded relevance trigger and time are (%s, %s), but notification on local NetworkAttachment %s makes real relevance trigger and time (%s, %s) (%f secs earlier).",
```
We have a mixture of decimal and hex formatting of VNIs.  I favor converging towards hex.

matte21 replied:
Switched to hex everywhere in 52413b4 . I plan a follow up where logs messages make it clearer that the VNI is in hex by wrapping the vni with "0x...h" .

MikeSpreitzer commented on staging/kos/pkg/controllers/connectionagent/connection_agent.go:
```diff
@@ -1051,7 +1175,12 @@ func (ca *ConnectionAgent) localAttachmentIsUpToDate(att *netv1a1.NetworkAttachm
 
-func (ca *ConnectionAgent) updateLocalAttachmentStatus(att *netv1a1.NetworkAttachment, macAddr, ifcName string, statusErrs sliceOfString, pcer *netv1a1.ExecReport) error {
+func (ca *ConnectionAgent) updateLocalAttachmentStatus(att *netv1a1.NetworkAttachment,
+	lastClientWriteTime time.Time,
```
I favor keeping `lastClientWrite` and `lastClientWriteTime` in their usual order, even though it makes this signature longer to write out.

matte21 replied:
Done in 52413b4 .

MikeSpreitzer commented on staging/kos/pkg/controllers/connectionagent/network_interface.go:
```diff
@@ -190,14 +190,18 @@ func (ca *ConnectionAgent) createLocalNetworkInterface(att *netv1a1.NetworkAttac
 	if att.Status.IfcName == "" {
-		ca.attachmentCreateToLocalIfcHistogram.Observe(tAfter.Sub(att.Writes.GetServerWriteTimeUnwrapped(netv1a1.NASectionSpec)).Seconds())
+		ca.lastClientWriteToLocalIfcHistograms.
+			With(prometheus.Labels{
```
BTW, I remembered that we could use `WithLabelValues` rather than `With`, making the code more concise.  I see the more concise usage in k/k.

matte21 replied:
Ah, yes, done in 52413b4 .

MikeSpreitzer commented on staging/kos/pkg/controllers/ipam/ipam.go:
```diff
@@ -85,14 +90,16 @@ type IPAMController struct {
-	// IPLock.CreationTimestamp - NetworkAttachment.CreationTimestamp
-	attachmentCreateToLockHistograms *prometheus.HistogramVec
+	// Seconds from creation of the last between a NetworkAttachment and its
+	// Subnet to creation of the NetworkAttachment's IPLock.
```
A more usual English rendering of that approach would be "Seconds from the later of the NetworkAttachment and Subnet creation times to ...".

How about "Seconds from the last relevant object creation to ..." ?

matte21 replied:
Reworded in 52413b4 .

MikeSpreitzer commented on staging/kos/pkg/controllers/ipam/ipam.go:
```diff
@@ -758,7 +764,22 @@ func (ctlr *IPAMController) pickAndLockAddress(ns, name string, att *netv1a1.Net
 			klog.V(4).Infof("Locked IP address %s for %s/%s=%s, lockName=%s, lockUID=%s, Status.IPv4 was %q", ipForStatus, ns, name, string(att.UID), lockName, string(ipl2.UID), att.Status.IPv4)
 			if len(att.Status.IPv4) == 0 {
-				ctlr.attachmentCreateToLockHistograms.With(prometheus.Labels{contentionLabel: strconv.FormatBool(prevAddressContention)}).Observe(ipl2.Writes.GetServerWriteTime(netv1a1.IPLockSectionSpec).Sub(att.Writes.GetServerWriteTimeUnwrapped(netv1a1.NASectionSpec)).Seconds())
+				naCreatTime := att.Writes.GetServerWriteTimeUnwrapped(netv1a1.NASectionSpec)
```
let's include the final "e" in "create".  As a rule.

matte21 replied:
Done in 52413b4 .

MikeSpreitzer commented on staging/kos/pkg/controllers/ipam/ipam.go:
```diff
@@ -758,7 +764,22 @@ func (ctlr *IPAMController) pickAndLockAddress(ns, name string, att *netv1a1.Net
 			if len(att.Status.IPv4) == 0 {
-				ctlr.attachmentCreateToLockHistograms.With(prometheus.Labels{contentionLabel: strconv.FormatBool(prevAddressContention)}).Observe(ipl2.Writes.GetServerWriteTime(netv1a1.IPLockSectionSpec).Sub(att.Writes.GetServerWriteTimeUnwrapped(netv1a1.NASectionSpec)).Seconds())
+				naCreatTime := att.Writes.GetServerWriteTimeUnwrapped(netv1a1.NASectionSpec)
+				if naCreatTime.After(subnetCreatTime) {
```
There is a lot of common code between the two branches of this `if` statement.  I think it would be better to use an `if` statement to set up two variables (the last client write identifier and time) and then use common code for the rest.

matte21 replied:
Done in 52413b4 .

MikeSpreitzer commented on staging/kos/pkg/controllers/ipam/ipam.go:
```diff
@@ -829,12 +851,33 @@ func (ctlr *IPAMController) updateNAStatus(ns, name string, att *netv1a1.Network
-		deltaS := att3.Writes.GetServerWriteTime(netv1a1.NASectionAddr).Sub(att.Writes.GetServerWriteTimeUnwrapped(netv1a1.NASectionSpec)).Seconds()
-		ctlr.attachmentCreateToAddressHistograms.With(prometheus.Labels{contentionLabel: strconv.FormatBool(nadat.addressContention)}).Observe(deltaS)
+		deltaSecs := "unknown"
+		if (subnetCreatTime != time.Time{}) {
```
Aha, I thought there were still problems with this histogram.  It was being updated in some inappropriate circumstances.  I think the right test is whether `len(statusErrs) == 0`, or equivalently, `ipForStatus != nil`.

matte21 replied:
Done in 52413b4 , but I was planning (and still do) a PR that tracks the "update without address" case as well via a label.

matte21 committed:
Various improvements suggested in code review

Follows a list of changes.

Rename: "creat" -> "create".

Access to Prometheus histograms and counters: from "With(...)" ->
"WithLabelValues(...)".

Refactorings to streamline code and reduce duplication.

Better wording of comments on Prometheus metrics.

Enforce consistent pass order of related function parameters.

Print VNI consistently in hex in logs and error messages.

Get rid of uninteresting Prometheus metrics.

MikeSpreitzer commented on staging/kos/pkg/controllers/connectionagent/connection_agent.go:
```diff
@@ -687,8 +736,35 @@ func (ca *ConnectionAgent) processQueueItem(attNSN k8stypes.NamespacedName, qlen
+	return
+}
+
+func getAttLastClientWriteInfo(att *netv1a1.NetworkAttachment, naLastClientWrVal, naSubnetLastClientWrVal string) (lastClientWr string, lastClientWrTime time.Time) {
```
The name of this func, "getAttLastClientWriteInfo", when compared with the preceding "getLastClientWriteInfo", differs only in the presence of "Att" --- which does not identify the difference between these two funcs.  Perhaps this one should be named something more distintive.  Perhaps something like "lastClientWriteInfoSwitch"?

matte21 replied:
The reasoning behind the name "getAttLastClientWriteInfo" is that it returns the last client write concerning an attachment, whereas the enclosing func returns the last client write among a group of writes that is related not (necessarily) just to the attachment being processed.

MikeSpreitzer replied:
Both are rooted at the NA, and even `getAttLastClientWriteInfo` looks further.
It is not just how far the func looks, but `getAttLastClientWriteInfo` takes parameters to use in characterizing the answer.

MikeSpreitzer commented:
We can improve that func name later.
